File: apps/api/src/star_itsm_api/schemas/ticket.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from star_itsm_api.schemas.attachment import AttachmentRead
from star_itsm_api.schemas.comment import CommentRead
from star_itsm_api.schemas.sub_cause import SubCauseRead
from star_itsm_api.schemas.ticket_activity import TicketActivityItemRead, TicketTimestampsRead
from star_itsm_api.schemas.ticket_intelligence import TicketIntelligenceRead
from star_itsm_api.schemas.ticket_routing import TicketRoutingRead
from star_itsm_api.services.cpr import assert_no_cpr_outside_field, validate_cpr
from star_itsm_api.services.ticket_tags import normalize_tags, validate_emoji


CLOSED_STATUSES = frozenset({"closed", "cancelled"})


class TicketSummaryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    ticket_number: str
    title: str
    status: str
    priority: str
    is_major: bool = False
    is_security_ticket: bool = False


class TicketRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    ticket_number: str
    title: str
    status: str
    priority: str
    ticket_type: str
    is_major: bool = False
    is_shared: bool = False
    is_security_ticket: bool = False
    parent_ticket_id: UUID | None = None
    parent: TicketSummaryRead | None = None
    child_count: int = 0
    sub_causes: list[SubCauseRead] = Field(default_factory=list)
    category_name_da: str | None = None
    subcategory_name_da: str | None = None
    assigned_team_id: UUID | None = None
    assigned_team_name: str | None = None
    assigned_user_name: str | None = None
    reporter_display_name: str | None = None
    response_due_at: datetime | None = None
    resolution_due_at: datetime | None = None
    sla_remaining_seconds: int | None = None
    sla_breached: bool = False
    created_at: datetime
    updated_at: datetime | None = None
    fault_displayed: bool = False
    tags: list[str] = Field(default_factory=list)
    emoji: str | None = None
    routing: TicketRoutingRead | None = None
    is_knowledge_article: bool = False
    knowledge_status: str | None = None
    knowledge_status_label_da: str | None = None
    knowledge_visibility: str | None = None
    knowledge_visibility_label_da: str | None = None


class TicketDetailRead(TicketRead):
    children: list[TicketSummaryRead] = Field(default_factory=list)
    related_major_tickets: list[TicketSummaryRead] = Field(default_factory=list)
    intelligence: TicketIntelligenceRead | None = None
    description: str
    category_id: UUID | None
    subcategory_id: UUID | None
    assigned_team_id: UUID | None
    assigned_team_name: str | None = None
    assigned_user_id: UUID | None = None
    assigned_user_name: str | None = None
    response_due_at: datetime | None
    resolution_due_at: datetime | None
    escalation_level: int
    assignment_reason: str | None = None
    fault_displayed: bool = False
    gdpr_consent: bool = False
    gdpr_consent_at: datetime | None = None
    subject_cpr: str | None = None
    attachments: list[AttachmentRead] = Field(default_factory=list)
    comments: list[CommentRead] = Field(default_factory=list)
    timestamps: TicketTimestampsRead
    activity: list[TicketActivityItemRead] = Field(default_factory=list)


class TicketCreate(BaseModel):
    ticket_type: Literal["service_request", "incident", "problem"] = "incident"
    title: str = Field(min_length=3, max_length=256)
    description: str = Field(min_length=10)
    priority: Literal["critical", "high", "medium", "low"] = "medium"
    category_id: UUID | None = None
    subcategory_id: UUID | None = None
    sub_cause_ids: list[UUID] = Field(default_factory=list)
    is_major: bool = False
    is_security_ticket: bool = False
    parent_ticket_id: UUID | None = None
    gdpr_consent: bool = False
    subject_cpr: str | None = Field(default=None, max_length=20)
    tags: list[str] = Field(default_factory=list, max_length=10)
    emoji: str | None = Field(default=None, max_length=16)
    intake_answers: dict[str, str] = Field(default_factory=dict, max_length=20)

    @field_validator("tags")
    @classmethod
    def normalize_create_tags(cls, value: list[str]) -> list[str]:
        return normalize_tags(value)

    @field_validator("emoji")
    @classmethod
    def validate_create_emoji(cls, value: str | None) -> str | None:
        return validate_emoji(value)

    @field_validator("subject_cpr")
    @classmethod
    def normalize_subject_cpr(cls, value: str | None) -> str | None:
        if value is None or not value.strip():
            return None
        return validate_cpr(value)

    @model_validator(mode="after")
    def privacy_fields(self) -> "TicketCreate":
        assert_no_cpr_outside_field(
            subject_cpr=self.subject_cpr,
            title=self.title,
            description=self.description,
        )
        if self.subject_cpr and not self.gdpr_consent:
            raise ValueError(
                "Du skal acceptere behandling af personoplysninger (GDPR) når du angiver CPR"
            )
        if not self.subject_cpr:
            self.gdpr_consent = False
        return self


class TicketParentUpdate(BaseModel):
    parent_ticket_id: UUID | None = None


class TicketRelatedMajorCreate(BaseModel):
    related_ticket_id: UUID


class TicketPriorityUpdate(BaseModel):
    priority: Literal["critical", "high", "medium", "low"]
    reason: str = Field(min_length=10, max_length=2000)


class TicketMetadataUpdate(BaseModel):
    is_major: bool | None = None
    is_security_ticket: bool | None = None
    parent_ticket_id: UUID | None = None
    sub_cause_ids: list[UUID] | None = None
    tags: list[str] | None = None
    emoji: str | None = Field(default=None, max_length=16)

    @field_validator("tags")
    @classmethod
    def normalize_metadata_tags(cls, value: list[str] | None) -> list[str] | None:
        if value is None:
            return None
        return normalize_tags(value)

    @field_validator("emoji")
    @classmethod
    def validate_metadata_emoji(cls, value: str | None) -> str | None:
        return validate_emoji(value)


class TicketStatusUpdate(BaseModel):
    status: Literal[
        "new",
        "assigned",
        "in_progress",
        "on_hold",
        "resolved",
        "closed",
        "cancelled",
    ]


class TicketAssignmentUpdate(BaseModel):
    assigned_team_id: UUID | None = None
    assigned_user_id: UUID | None = None
    assignment_reason: str | None = Field(default=None, max_length=2000)
    fault_displayed: bool | None = None

File: apps/web/src/types/ticket.ts
import type { Attachment } from "@/types/attachment";
import type { Comment } from "@/types/comment";
import type { SubCause } from "@/types/sub-cause";
import type { TicketActivityItem, TicketTimestamps } from "@/types/ticket-activity";

export interface TicketSummary {
  id: string;
  ticket_number: string;
  title: string;
  status: string;
  priority: string;
  is_major: boolean;
}

export interface Ticket {
  id: string;
  ticket_number: string;
  title: string;
  description?: string;
  status: string;
  priority: string;
  ticket_type: string;
  is_major: boolean;
  is_shared?: boolean;
  is_security_ticket?: boolean;
  parent_ticket_id?: string | null;
  parent?: TicketSummary | null;
  child_count?: number;
  sub_causes: SubCause[];
  category_name_da?: string | null;
  subcategory_name_da?: string | null;
  assigned_team_id?: string | null;
  assigned_team_name?: string | null;
  assigned_user_id?: string | null;
  assigned_user_name?: string | null;
  fault_displayed?: boolean;
  assignment_reason?: string | null;
  reporter_display_name?: string | null;
  response_due_at?: string | null;
  resolution_due_at?: string | null;
  sla_remaining_seconds?: number | null;
  sla_breached?: boolean;
  created_at: string;
  updated_at?: string | null;
  tags?: string[];
  emoji?: string | null;
  routing?: TicketRouting | null;
  is_knowledge_article?: boolean;
  knowledge_status?: string | null;
  knowledge_status_label_da?: string | null;
  knowledge_visibility?: string | null;
  knowledge_visibility_label_da?: string | null;
  /** DB: portal | email | api | phone | chat | knowledge */
  source?: string;
  source_label_da?: string;
}

export interface TicketIntake {
  answers: Record<string, string>;
}

export interface TicketRouting {
  completeness_score: number;
  routing_ready: boolean;
  missing_fields_da: string[];
  intake: TicketIntake;
  suggested_team_id?: string | null;
  suggested_team_name?: string | null;
  routing_confidence?: number | null;
  routing_reason_da?: string | null;
  computed_priority: string;
  computed_priority_label_da: string;
  computed_priority_reasons_da: string[];
}

export interface TicketIntelligence {
  semantic_topics: string[];
  ease_score: number | null;
  ease_label_da: string | null;
  complexity_score: number | null;
  complexity_label_da: string | null;
  llm_summary: string | null;
  handling_hints: string[];
  source: string | null;
  updated_at: string | null;
}

export interface TicketEmail {
  id: string;
  direction: "inbound" | "outbound";
  subject?: string | null;
  from_email?: string | null;
  to_email?: string | null;
  body_text?: string | null;
  received_at: string;
}

export interface TicketLlmContext {
  schema_version: string;
  ticket_id: string;
  ticket_number: string;
  intelligence: TicketIntelligence;
  semantic_bundle: {
    title: string;
    description: string;
    tags: string[];
    emoji: string | null;
    category_name_da: string | null;
    subcategory_name_da: string | null;
    sub_cause_names_da: string[];
    combined_text: string;
  };
  operational: Record<string, unknown>;
  routing?: TicketRouting | null;
  prompt_snippet_da: string;
  evaluation_rubric_da: string;
}

export interface TicketDetail extends Ticket {
  children?: TicketSummary[];
  related_major_tickets?: TicketSummary[];
  intelligence?: TicketIntelligence | null;
  description: string; // required on detail
  category_id: string | null;
  subcategory_id: string | null;
  assigned_team_id: string | null;
  assigned_team_name: string | null;
  assigned_user_id: string | null;
  assigned_user_name: string | null;
  response_due_at: string | null;
  resolution_due_at: string | null;
  escalation_level: number;
  gdpr_consent: boolean;
  gdpr_consent_at: string | null;
  subject_cpr: string | null;
  attachments: Attachment[];
  comments: Comment[];
  ticket_emails?: TicketEmail[];
  linked_gmail_email?: string | null;
  timestamps?: TicketTimestamps;
  activity?: TicketActivityItem[];
}

export interface TicketCreateInput {
  ticket_type: "service_request" | "incident" | "problem";
  title: string;
  description: string;
  priority: "critical" | "high" | "medium" | "low";
  category_id?: string | null;
  subcategory_id?: string | null;
  sub_cause_ids?: string[];
  is_major?: boolean;
  is_security_ticket?: boolean;
  parent_ticket_id?: string | null;
  gdpr_consent: boolean;
  subject_cpr?: string | null;
  tags?: string[];
  emoji?: string | null;
  intake_answers?: Record<string, string>;
  /** Kun agenter — sendes til API; slutbrugere ignorerer feltet (portal). */
  source?: "portal" | "email" | "phone" | "chat";
}
